### ISSUE TITLE
[Urgent Help needed] why can Poisson/NegativeBinomial regression work with Standardized count data (including negative values) using bambi? 

### ISSUE THREAD

**Issue by SGongS** [bambinos/bambi#443]
Hi, I have the count data (number of trips between origin and destination pairs) and try to build statistical models for the count data using bambi Poisson and NegativeBinomial regression model. However, when I test the model estimates against ground truth based on metrics like mean absolute error or mean squared error, the errors are very large (much much bigger than 1000). Even though the summary tables and trace plots show that I have derived parameters with stationary traces, I couldn't believe that the bambi models with the NUTS sampler worked correctly for my over-dispersed counts. I have read an example that applied Bayesian Inference with PyMC3 and Bambi(https://towardsdatascience.com/applied-bayesian-inference-with-pymc3-and-bambi-pt-3-d4bfb3211509) and got the idea to first standardize the counting measure, then implement Poisson and NegativeBinomial regression using standardized count (including negative values) with bambi. During the fitting, there are no warnings or errors. When I test the model estimates against the standardized count values, I derive quite reasonable mean absolute error or mean squared error (at the level of 0.1 or 0.01 something). I don't understand why bambi Poisson and NegativeBinomial models can work with negative values or can work better with negative values than original non-negative count data, as the logarithm of a negative value is impossible in the real numbers. Can anyone help?

**Comment by canyon289** [comment]
@SGongS, thanks for all the information. Can you share a minimal reproducible notebook? I'm not an expert in regression so I'm not sure I'll be able to answer your question. Seeing an example would go a long way helping us help you

**Comment by SGongS** [comment]
@canyon289 Thank you for your answer. Here is a minimal reporducible notebook with an attached dataset
import bambi as bmb
[shared_with_bambi_github.csv](https://github.com/bambinos/bambi/files/7892248/shared_with_bambi_github.csv)
import bambi as bmb
import pandas as pd
import statsmodels.api as sm
import arviz as az
import matplotlib.pyplot as plt
import numpy as np
import geopandas as gpd
import patsy as pt
import scipy.stats as ss
import seaborn as sns
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error

data = pd.read_csv('outputs/shared_with_bambi_github.csv')
#prepare training and testing dataset
mask = np.random.rand(len(data)) < 0.8
df_train = data[mask]
df_test = data[~mask]
print('Training data set length='+str(len(df_train)))
print('Testing data set length='+str(len(df_test)))M

if __name__ == '__main__':
    modelMon_bmb_nb_glm = bmb.Model("tijMon ~ p1 + p2 + dMP", df_train, family="negativebinomial")
    fittedMon_nb_glm = modelMon_bmb_nb_glm.fit(draws=500, tune=2000, target_accept=.95, chains=2,cores=4,init='adapt_diag')
    modelMon_bmb_poisson_glm = bmb.Model("tijMon ~ p1 + p2 + dMP", df_train,  family="poisson")
    fittedMon_poisson_glm = modelMon_bmb_poisson_glm.fit(draws=500, tune=2000, target_accept=.95, chains=2,cores=4,init='adapt_diag')

posterior_predictive_hier_glm = modelMon_bmb_nb_glm.predict(fittedMon_nb_glm, kind="pps", data=df_test)
y_pred = fittedMon_nb_glm['posterior_predictive']['tijMon'].values.mean(axis=0).mean(axis=0)
print("Mean Absolute Error",mean_absolute_error(df_test['tijMon'].values,y_pred))
print("Mean Squared Error",mean_squared_error(df_test['tijMon'].values,y_pred))
print("Root Mean Squared Error",np.sqrt(mean_squared_error(df_test['tijMon'].values,y_pred)))
#standardization
varnames_o = ['tijMon','p1','p2','dMP']
train = df_train.copy()
test = df_test.copy()
standard_scaler = StandardScaler()
train = standard_scaler.fit_transform(train)
test = standard_scaler.transform(test)
train = pd.DataFrame(train, columns=varnames_o)
test = pd.DataFrame(test, columns=varnames_o)

if __name__ == '__main__':
    modelMon_bmb_nb_glm2 = bmb.Model("tijMon ~ p1 + p2 + dMP", train, family="negativebinomial")
    fittedMon_nb_glm2 = modelMon_bmb_nb_glm2.fit(500, tune=2000, target_accept=.95, chains=2,cores=4,init='adapt_diag')
    modelMon_bmb_poisson_glm2 = bmb.Model("tijMon ~ p1 + p2 + dMP", train,  family="poisson")
    fittedMon_poisson_glm2= modelMon_bmb_poisson_glm2.fit(500, tune=2000, target_accept=.95, chains=2,cores=4,init='adapt_diag')

posterior_predictive_hier_glm_ = modelMon_bmb_nb_glm2.predict(fittedMon_nb_glm2, kind="pps", data=test)
y_pred = fittedMon_nb_glm2['posterior_predictive']['tijMon'].values.mean(axis=0).mean(axis=0)
print("Mean Absolute Error",mean_absolute_error(test['tijMon'].values,y_pred))
print("Mean Squared Error",mean_squared_error(test['tijMon'].values,y_pred))
print("Root Mean Squared Error",np.sqrt(mean_squared_error(test['tijMon'].values,y_pred)))

**Comment by tomicapretto** [comment]
Hi @SGongS! Let me add some information

1. It's really weird the model works for non-integer and negative response values.  But this is not a problem with Bambi itself, but a problem with PyMC3 (note I say PyMC3 because Bambi currently uses PyMC3 and not PyMC v4). This is the code that reproduces the code in the underlying model in Bambi


```python
import pymc3 as pm
import theano.tensor as tt

# This 'train' is the standardized one
X = train[["p1", "p2", "dMP"]].values
y = train["tijMon"].values

with pm.Model() as model:
    b0 = pm.Normal("b0", mu=0, sigma=10)
    betas = pm.Normal("betas", mu=0, sigma=10, shape=3)
    
    nu = b0 + tt.dot(X, betas)
    mu = tt.exp(nu)
    alpha = pm.HalfCauchy("alpha", beta=1)
    
    obs = pm.NegativeBinomial("obs", mu=mu, alpha=alpha, observed=y)   
    
    idata = pm.sample(return_inferencedata=True)
```

As you can see the model samples well, even when the response `y` is non-integer and has some negative values. @ricardoV94 is this something you are aware of? Do we have the same problem in PyMC v4?

2. Measurements such as Mean Squared Error, Mean Absolute Error, etc. depend on the magnitude of the response. If your response has smaller values, it's expected to have smaller error values. That should explain the different results you are obtaining, at least partially.

3. 

Bambi provides support to standardize predictors in the model formula itself without having to do that manually as a pre-processing step with the function `scale()`. In this case, because of the magnitude of your predictors, it makes sense to standardize so their values are expressed in smaller units. 

You can also pass a custom prior for the "alpha" auxiliary parameter in the Negative Binomial distribution. I noticed the posterior is very small and the default HalfCauchy prior may not be a good idea. 

Something along these lines should work better:


```python
priors = {"alpha": bmb.Prior("Exponential", lam=1)}
model = bmb.Model("tijMon ~ scale(p1) + scale(p2) + scale(dMP)", df_train, family="negativebinomial", priors=priors)
idata = model.fit(target_accept=0.9)
```

**Comment by ricardoV94** [comment]
We don't have checks for discrete data, and all the logp function in `NegativeBinomial` work for continuous values. The `logp` should be `-inf` for negative values though... Can someone provide a MWE with fake data that can be run directly?

**Comment by SGongS** [comment]
Hi @tomicapretto thank you for the added information. Regarding your point 3, I have tested by myself as well. However, I didn't see the improvement. May I ask based on which metric you say with customized prior here works better? Actually, the negative binomial model here works really bad for these response values compared with Poisson model (based on r_value distribution between predicted and observed), which is another very wired thing, for me as the data contains a lot of zero. 
> You can also pass a custom prior for the "alpha" auxiliary parameter in the Negative Binomial distribution. I noticed the posterior is very small and the default HalfCauchy prior may not be a good idea.
> 
> Something along these lines should work better:
> 
> ```python
> priors = {"alpha": bmb.Prior("Exponential", lam=1)}
> model = bmb.Model("tijMon ~ scale(p1) + scale(p2) + scale(dMP)", df_train, family="negativebinomial", priors=priors)
> idata = model.fit(target_accept=0.9)
> ```
a plot example for negative binomial in bambi/pymc3
![image](https://user-images.githubusercontent.com/62221787/150558178-856a2921-b91a-4042-9752-16436e1b0ad6.png)

**Comment by SGongS** [comment]
@tomicapretto do you have any suggestions for using zero-inflated Poisson in Bambi/pymc3 for this example?

**Comment by ricardoV94** [comment]
> @ricardoV94 thank you for your explanation. If this is the truth in pymc3 (Note that the observed values may be converted to integer behind the scenes, and values like -0.9 will become 0.), can I say that the outputs of either Poisson or negative binomial in pymc3 using non-integer or negative response values should be douted a lot, as the response values will be totally changed from the original one? Or can I say the count data should not be standardized or scaled if we would like to fit it in the bayesian Poisson or negative binomial model?

They should be doubted, as they are not valid values

**Comment by tomicapretto** [comment]
I think the original question is already answered
